luaui/Widgets/gui_top_bar.lua

```diff
@@ -2054,22 +3224,125 @@ function shutdown()
+	-- TODO: can we get rid of cmdlist? Moving should count as non-idle for at least commanders, and perhaps other units too.
+	local unitExists = false
+	local active = false
+	local builtUnitDefID = nil
+	local mayBeBuilding = false -- will remain false even if builtUnitDefID is non-nil
+	local guardedUnitID = nil -- returned in case we want to check for cycles of idle builders
+
+	local commands = nil
+	-- We only need one command to figure out if the unit is active.
+	-- TODO: Try GetUnitCurrentCommand
+	if UnitDefs[unitDefID].isFactory then
+		commands = Spring.GetFactoryCommands(unitID, 1)
+	else
+		commands = Spring.GetUnitCommands(unitID, -1)
+	end
+
+	if commands then
+		unitExists = true
+		for i = 1, #commands do
+			for _, relevantCMD in ipairs(cmdList) do
+				if commands[i].id == relevantCMD or commands[i].id < 0 then
+					active = true
+					if commands[i].id < 0 and not builtUnitDefID then
+						-- We're building something. A negative ID is the unitdef ID of what's being built.
+						builtUnitDefID = -commands[i].id
+						--Spring.Echo("unit " .. unitID .. " is building unitdef " .. builtUnitDefID)
+					elseif commands[i].id == CMD.GUARD then
+						mayBeBuilding = true -- building can be caused by a guard command
+						guardedUnitID = commands[i].params[1]
+					elseif commands[i].id == CMD.REPAIR then
+						mayBeBuilding = true -- building can be caused by a repair command
+					end
+				end
+			end
+		end
+	end
+
+	return unitExists, active, builtUnitDefID, mayBeBuilding, guardedUnitID
 end
 
-function widget:SetConfigData(data)
-	if data.autoHideButtons ~= nil then
-		autoHideButtons = data.autoHideButtons
+function pid(Kp, Ki, Kd, dt, prevIntegral, prevError, setPoint, measuredValue)
+	if dt <= 0 then
+		return measuredValue, prevIntegral, prevError
+	end
+	-- setpoint: instantaneous observed value
+	-- measuredValue: position of displayed indicator
+	local error = setPoint - measuredValue
+	local proportional = error;
+	local integral = prevIntegral + error * dt
+	local derivative = (error - prevError) / dt
+	local output = Kp * proportional + Ki * integral + Kd * derivative
+	return output, integral, error
+end
+
+function TrackUnit(unitID, unitDefID, unitTeam) --needed for exact calculations
+	if (myTeamID == unitTeam) then
+		local unitDef = UnitDefs[unitDefID]
+		local _, _, _, _, build = Spring.GetUnitHealth(unitID) --is it finished?
+		local isBuilt = build >= 1
+		if isBuilt and unitDef then
+			if unitDef.buildSpeed and unitDef.buildSpeed > 0 and unitDef.canAssist or UnitDefs[unitDefID].isFactory then
+				trackedNum = trackedNum + 1
+				if not trackedBuilders[unitID] then				-- We may have already tracked this unit when it started being built. No need to add its BP again.
+					BP[4] = BP[4] + unitDef.buildSpeed -- BP[4] ^= totalAvailableBP
+				end
+				trackedBuilders[unitID] = { unitDef.buildSpeed, isBuilt, unitDefID, unitTeam }
```

Comment:
Can we give these values keys, for readability?

Reply:
I'd love to.

I strugle with this one, though. 'BP' gets used like 'energy' and 'metal' And they are used like BP = {0, 0 ,1} and so on
I haven't found a way to give these values keys, yet.

PLS help :)

Reply:
done where possible